### PR TITLE
Rework bundler and bring back support to es2017 

### PR DESCRIPTION
**Pull request by belgattitude**
## Purpose

Rework the bundler config to ensure emitted file are valid es2017 files. Will fix https://github.com/imbhargav5/rooks/issues/1362

## Approach

Tried to keep on rollup, but easily extensible for esbuild

## How to try it out

```basjh
cd ./packages/rooks
yarn build
yarn check:dist
yarn check:size
```

## Checks

Impact on size: negligible imo

image

## Tasks

- [x] Emit es2017 thx typescript (could work with esbuild too)
- [x] Add module export that might help some bundlers (no changes)
- [x] Removed unused files (ie: jest.setup.ts, test files)
- [x] Removed declaration maps  for dist (they are only useful in monorepo ?)
- [x] Removed dual types publishing (kept ./types/index.d.ts, removed per-file declarations)
- [x] Didn't understand what was the purpose of index.standalone, need to clear this
- [x] Merge rollup configs, and allow to build umd as well   - **To test it out**
- [x] Post-cleanup: remove obsoleted deps...

### CODE REVIEW

File: packages/rooks/package.json [hunk 1]
```diff
@@ -7,6 +7,15 @@
   "unpkg": "dist/umd/rooks.umd.js",
   "types": "dist/types/index.d.ts",
   "typings": "dist/types/index.d.ts",
+  "exports": {
+    ".": {
+      "types": "./dist/types/index.d.js",
+      "module": "./dist/esm/index.js",
```
Comment (belgattitude) [inline packages/rooks/package.json:13]: Kept the `module` for webpack support (it's not part of the final spec). Not sure if recent versions needs that.

Didn't add the `browser` field (umd), cause nextjs 12.x will pick it up and defeat tree-shakability (browser bundle is one file)

File: packages/rooks/rollup.config.ts [hunk 2]
```diff
@@ -14,6 +14,7 @@ const commonPlugins = [
   sourceMaps(),
   json(),
   babel({
+    babelHelpers: "bundled",
```
Comment (belgattitude) [inline packages/rooks/rollup.config.ts:17]: removed a babel warning

File: packages/rooks/rollup.future.config.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    plugins: [
+      tsPathPlugin,
+      dts({
+        compilerOptions: {
+          sourceMap: false,
+        },
+      }),
+    ],
+  },
+  // @todo UMD from rollup.config.ts
```
Comment (belgattitude) [inline packages/rooks/rollup.future.config.ts:67]: I'll merge from the existing rollup config

File: packages/rooks/rollup.future.config.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+const external = [
+  ...Object.keys(pkg.dependencies),
+  ...Object.keys(pkg.peerDependencies || {}),
+];
+
+const distDir = "./dist/future";
+
+const tsPathPlugin = typescriptPaths({
+  preserveExtensions: true,
+  tsConfigPath: "./tsconfig.json",
```
Comment (belgattitude) [inline packages/rooks/rollup.future.config.ts:17]: typescriptPaths plugin does not support extend from tsconfig.build.json. So I use the paths from there. Note that it makes use remove composite: true.

**Comment by belgattitude** [comment]
Hey @imbhargav5. It's just a start.

But I'd like to know

- What's the purpose of publishing index.standalone.js ? Is it important ?
- Are the declaration sourcempas and sourcemaps important ?

I'm not totally sure about this.

Happy week-end.

File: packages/rooks/tsconfig.json [hunk 5]
```diff
@@ -9,7 +9,6 @@
     "module": "esnext",
     "jsx": "react",
     "target": "esnext",
-    "composite": true,
```
Comment (belgattitude) [inline packages/rooks/tsconfig.json:12]: I had to remove this cause I will make the rollup-plugin-typescript-paths fail. See above comment.

I feel there's no need to set it, cause I haven't seen project references support

**Comment by imbhargav5** [comment]
Going to look at this tonight!

File: packages/rooks/package.json (hunk 1 above)
Comment (imbhargav5) [reply packages/rooks/package.json:13]: @belgattitude Could you help me with any reading material for me to fully understand what has changed in the `exports` field? It changes so often I lost track of it. Just want to make sure I understand all the changes we are making and we don't accidentally break anything.

File: packages/rooks/tsconfig.json (hunk 5 above)
Comment (imbhargav5) [reply packages/rooks/tsconfig.json:12]: Got it. If we don't need it, let's scrap. 👍🏽

**Comment by imbhargav5** [comment]
@belgattitude `index.standalone` is an artefact from a year ago and we don't need that anymore. Thanks for pointing that out. We can remove it. 😄

File: packages/rooks/package.json (hunk 1 above)
Comment (belgattitude) [reply packages/rooks/package.json:13]: Most up-to-date I found out is

https://github.com/frehner/modern-guide-to-packaging-js-library#define-your-exports

**Comment by belgattitude** [comment]
@imbhargav5 

If all looks good to you, my plan is to replace the build with build-future and the new rollup config. I'll undraft when done.

My last concern, I saw a drop in size in the umd bundle. I guess is that the *.spec.ts aren't treated anymore. But I'd like another look.

File: packages/rooks/package.json (hunk 1 above)
Comment (belgattitude) [reply packages/rooks/package.json:13]: If you feel unsure I can cherry-pick this and create another PR.

**Comment by imbhargav5** [comment]
Will take a look.

File: packages/rooks/package.json (hunk 1 above)
Comment (imbhargav5) [reply packages/rooks/package.json:13]: @belgattitude Thanks for this! I will read this and get back. I will take a good look at this PR tomorrow.

Comment (imbhargav5) [reply packages/rooks/package.json:13]: TIL learnt a lot from that link today. Thanks!

**Comment by imbhargav5** [comment]
@belgattitude  Looks like a great start. I am getting errors after I run `yarn run build:future && yarn run check:dist`. Is that expected?

**Comment by belgattitude** [comment]
Yes it will check the ./dist/* folder, not the ./dist/future :smile: .  I'll P/R tomorrow.

**Comment by belgattitude** [comment]
Hey @imbhargav5 I've just finalized the thing.  

Added 
- a (patch) changeset
- a probably useful 'typecheck' step (no on CI though), that can pave the way for using esbuild.

Entered pre canary mode, so If you don't mind  I'd like to test a canary version release first.

**Comment by belgattitude** [comment]
Found a little time and fixed conflicts.

Ported the check-size in #1365 

Good news, the size increase is very minimal

image

The build passed from 20s to 15s on my laptop. Imo it can be merged, let me know if you have questions.

Thanks for this beautiful lib

File: packages/rooks/.size-limit.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+// @ts-check
+
+// Regularly update this max size when new hooks are added (keep a threshold)
+const fullBundleMaxSize = "12KB";
```
Comment (belgattitude) [inline packages/rooks/.size-limit.js:4]: @imbhargav5 just to mention, if new hooks are added and the total size of lib grow, this should be increased.

On the other side, it's totally possible to keep only the `"ESM Webpack ({ useDebounce })"` (see below), that's actually the important check

File: packages/rooks/package.json [hunk 7]
```diff
@@ -7,6 +7,15 @@
   "unpkg": "dist/umd/rooks.umd.js",
   "types": "dist/types/index.d.ts",
   "typings": "dist/types/index.d.ts",
+  "exports": {
+    ".": {
+      "types": "./dist/types/index.d.js",
```
Comment (imbhargav5) [inline packages/rooks/package.json:12]:
```suggestion
      "types": "./dist/types/index.d.ts",
```

File: packages/rooks/.size-limit.js (hunk 6 above)
Comment (imbhargav5) [reply packages/rooks/.size-limit.js:4]: @belgattitude Ah. I think we could write a small script to autopopulate new hooks into the size-limit file when new hooks are added and keep the limit at 2kb. What do you think?

**Comment by imbhargav5** [comment]
This is such a massive improvement @belgattitude ! I just want to keep this in `pre` mode for a bit for a couple of days. Can we re-enter pre mode again?

**Comment by imbhargav5** [comment]
@belgattitude I continued your work here. https://github.com/imbhargav5/rooks/pull/1408

File: packages/rooks/.size-limit.js (hunk 6 above)
Comment (belgattitude) [reply packages/rooks/.size-limit.js:4]: Nice idea

**Pull request closed without merging**